ksqldb-engine/src/main/java/io/confluent/ksql/internal/StorageUtilizationMetrics.java

```diff
@@ -0,0 +1,254 @@
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.internal;
+
+import com.google.common.collect.ImmutableMap;
+import io.confluent.ksql.metrics.MetricCollectors;
+import java.io.File;
+import java.math.BigInteger;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.KafkaMetric;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.streams.StreamsConfig;
+
+public class StorageUtilizationMetrics implements MetricsReporter {
+  private static final String METRIC_GROUP = "ksqldb-utilization";
+
+  private final Map<String, Map<String, TaskStorageMetric>> metricsSeen;
+  private final Metrics metricRegistry;
+  static AtomicBoolean registeredNodeMetrics = new AtomicBoolean(false);
+
+  public StorageUtilizationMetrics() {
+    this(MetricCollectors.getMetrics());
+  }
+
+  public StorageUtilizationMetrics(final Metrics metricRegistry) {
+    this.metricsSeen = new HashMap<>();
+    this.metricRegistry = metricRegistry;
+  }
+
+  @Override
+  public void init(final List<KafkaMetric> list) {
+  }
+
+  @Override
+  public void configure(final Map<String, ?> map) {
+    final String dir;
+    if (map.containsKey(StreamsConfig.STATE_DIR_CONFIG)) {
+      dir = map.get(StreamsConfig.STATE_DIR_CONFIG).toString();
+    } else {
+      dir =  StreamsConfig
+        .configDef()
+        .defaultValues()
+        .get(StreamsConfig.STATE_DIR_CONFIG)
+        .toString();
+    }
+    configureShared(new File(dir), this.metricRegistry);
+  }
+
+  private static void configureShared(final File baseDir, final Metrics metricRegistry) {
+    if (registeredNodeMetrics.getAndSet(true)) {
+      return;
+    }
+    final MetricName nodeAvailable =
+        metricRegistry.metricName("node-storage-available", METRIC_GROUP);
+    final MetricName nodeTotal =
+        metricRegistry.metricName("node-storage-total", METRIC_GROUP);
+    final MetricName nodeUsed =
+        metricRegistry.metricName("node-storage-used", METRIC_GROUP);
```

Comment:
@rodesai I can't remember what we landed on for node metrics - IIRC calculated storage used by doing (f.getTotalSpace - f.getFreeSpace) isn't actually accurate for what we want. Given that, do we just want to report f.getFreeSpace? Or f.getFreeSpace and f.getTotalSpace?

Reply:
Additionally for naming these + the other storage metrics, do we want to add a unit suffix? eg `_bytes`

Reply:
> IIRC calculated storage used by doing (f.getTotalSpace - f.getFreeSpace) isn't actually accurate for what we want. Given that, do we just want to report f.getFreeSpace? Or f.getFreeSpace and f.getTotalSpace?

I thought we determined it was probably accurate, and the disparity we saw was very small (18MB) and probably explained by the workload continuing to run between the time you printed the metrics and the time you looked at `df`. So we decided to get a PR out and we can test on a cloud instance rather than on our macbooks.

And yeah a `_bytes` suffix makes sense - good call